tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,763 @@
+		}
+		return
+	}
+
+	pos := 0
+	if t.dimWeights != nil {
+		for i, w := range t.dimWeights {
+			pos += d[i] * w
+		}
+	} else {
+		t.dimWeights = make([]int, len(d))
+		pos = d[len(d)-1]
+		if pos >= t.Dim(len(d)-1) {
+			err = &ErrIndexOutOfRange{
+				dim:       len(d) - 1,
+				index:     pos,
+				dimsRange: t.Dim(len(d) - 1),
+			}
+			return
+		}
+		t.dimWeights[len(d)-1] = 1
+
+		lastWeight := 0
+		for i := len(d) - 2; i >= 0; i-- {
+			lastWeight += t.Dim(i + 1)
+			t.dimWeights[i] = lastWeight
+			pos += d[i] * lastWeight
+
+			if d[i] >= t.Dim(i) {
+				err = &ErrIndexOutOfRange{
+					dim:       i,
+					index:     pos,
+					dimsRange: t.Dim(i),
+				}
+				return
+			}
+		}
+	}
+
+	switch t.DataType() {
+	case DtFloat:
+		vals, _ := t.AsFloat32()
+		val = vals[pos]
+	case DtDouble:
+		vals, _ := t.AsFloat64()
+		val = vals[pos]
+	case DtInt8, DtInt16, DtInt32, DtUint8:
+		vals, _ := t.AsInt32()
+		val = vals[pos]
+	case DtInt64:
+		vals, _ := t.AsInt64()
+		val = vals[pos]
+	case DtBool:
+		vals, _ := t.AsBool()
+		val = vals[pos]
+	case DtString:
+		vals, _ := t.AsStr()
+		val = vals[pos]
+	default:
+		err = &ErrTensorTypeNotSupported{
+			tensotType: t.DataType(),
+		}
+		return
+	}
+
+	return
+}
+
+// setCMemAsAlreadyRelease The C allocated memory was already released from C.
+func (t *Tensor) setCMemAsAlreadyRelease() {
+	t.memReleased = true
+}
+
+// FreeAllocMem Method used telease the C allocated memory for this tensor.
+func (t *Tensor) FreeAllocMem() {
+	// We can't clean the tensor here in case of it had been  used as in
+	// input parameter since on tensorflow/core/client/tensor_c_api.cc the
+	// function TF_Run_Helper is cleaning the input tensors after every
+	// execution what can cause a a double free or corruption error in C++
+	// since there is no way to determine if a tensor had been previously
+	// cleaned.
+	if !t.memReleased {
+		TF_DeleteTensor(t.tensor)
+	}
+}
+
+// ErrInvalidTensorType The data type of the tensor is not compatible
+// with the expected data type on this function.
+type ErrInvalidTensorType struct {
+	tensorType   DataType
+	expectedType DataType
+}
+
+func (e *ErrInvalidTensorType) Error() string {
+	return fmt.Sprintf("Invalid tensor data type, tensor data type: '%s', required data type: '%s'", e.tensorType, e.expectedType)
+}
+
+// ErrTensorTypeNotSupported The tensor type is still not supported.
+type ErrTensorTypeNotSupported struct {
+	tensotType DataType
+}
+
+func (e *ErrTensorTypeNotSupported) Error() string {
+	return fmt.Sprintf("The tensor data type '%s' is still not supported", e.tensotType)
+}
+
+// ErrDimsOutOfTensorRange The number of specified dimensions doesn't
+// match with the tensor dimensions.
+type ErrDimsOutOfTensorRange struct {
+	tensorDim int
+	specDims  int
+}
+
+func (e *ErrDimsOutOfTensorRange) Error() string {
+	return fmt.Sprintf("The number of specified dimensions (%d) doesn't match with the tensor dimensions (%d)", e.specDims, e.tensorDim)
```

Comment:
s/number of specified dimensions/specified number of dimensions/
s/doesn't match with/doesn't match/